simplemma/language_detector.py

```diff
@@ -1,128 +1,179 @@
     dictionary_factory: DictionaryFactory = DictionaryFactory(),
     token_sampler: TokenSampler = TokenSampler(),
 ) -> float:
     """Determine which proportion of the text is in the target language(s)."""
-    total = 0
-    in_target = 0
-    dictionaries = dictionary_factory.get_dictionaries(lang)
-    for token in token_sampler.sample_tokens(text):
-        total += 1
-        for lang_code, lang_dictionary in dictionaries.items():
-            candidate = _return_lemma(
-                token, lang_dictionary, greedy=True, lang=lang_code
-            )
-            if candidate is not None:
-                in_target += 1
-                break
-    if total > 0 and in_target > 0:
-        return in_target / total
-    return 0
+    return LanguageDetector(dictionary_factory).detect_text_proportion_in_languages(
+        text, lang, greedy, token_sampler
+    )
 
 
-def _return_default() -> List[Tuple[str, float]]:
-    # todo: None if 'unk'?
-    return [("unk", 1)]
-
-
-def lang_detector(
+def _convert_results_to_sorted_list(
+    results: Dict[str, float]
+) -> List[Tuple[str, float]]:
+    list_results: List[Tuple[str, float]] = sorted(
+        results.items(), key=itemgetter(1), reverse=True
+    )
+    # switch unknown to the end
+    for i, item in enumerate(list_results):
+        if item[0] == "unk":
+            pair = list_results.pop(i)
+            list_results.append(pair)
+    return list_results
+
+
+def langdetect(
     text: str,
     lang: Optional[Union[str, Tuple[str, ...]]] = None,
     greedy: bool = False,
     dictionary_factory: DictionaryFactory = DictionaryFactory(),
-    token_sampler: TokenSampler = TokenSampler(),
-    backup_sampler: TokenSampler = RelaxedTokenSampler(),
+    token_samplers: List[TokenSampler] = [
+        TokenSampler(),
+        TokenSampler(
+            tokenizer=Tokenizer(RELAXED_SPLIT_INPUT),
+            max_tokens=1000,
+            capitalized_threshold=0,
+        ),
+    ],
 ) -> List[Tuple[str, float]]:
-    """Determine which proportion of the text is in the target language(s).
-    Perform a first run and further discriminate between the results if necessary."""
-    myresults = {}  # Dict[str, float]
-    tokens = token_sampler.sample_tokens(text)
-    total_tokens = len(tokens)
-    if total_tokens == 0:
-        return _return_default()
-    # iterate
-    dictionaries = dictionary_factory.get_dictionaries(lang)
-    for lang_code, lang_dictionary in dictionaries.items():
-        in_target = 0
+    """Determine which proportion of the text is in the target language(s)."""
+    for token_sampler in token_samplers:
+        results = LanguageDetector(
+            dictionary_factory
+        ).detect_text_proportion_in_each_language(text, lang, greedy, token_sampler)
+        list_results = _convert_results_to_sorted_list(results)
+
+        # post-processing
+        if len(list_results) == 1 or list_results[0][1] != list_results[1][1]:
+            return list_results
```

Comment:
I think `break` would be better here from a syntax perspective.

Reply:
I personally like early returns just as I like guard clauses.

Things like break are more likely to introduce issues if new code is added after the for loop.

But I understand that this is a matter of personal preference. So, I can switch to `break` if you prefer that.

Reply:
I think this code shouldn't move much so we could use break.

Reply:
Could you please make the change?

Reply:
I read a bit more about this (early exit vs break), we can keep it that way.